Delete assets on media removal

xaben opened the pull request:
Another similar solution for #504.

Ensures all media files are deleted on media removal.

@rande Hope this gets merged, as at the moment both the reference and the thumbnails are left in place after media removal.

OskarStark commented:
@xaben i really like this PR and want to merge this. Can you please rebase your PR and ping me afterwards?

this would be :+1:

xaben commented:
@OskarStark I can rebase it tonight if needed. The issue that this PR targeted (#504) was kind of fixed by #696 . Still I think my solution is a little better as it actually deletes the media inside the PostRemove handle and not in PreRemove.

Please check the issue and tell me if you want a rebase or shall we close this PR.

Thanks.

core23 commented on Model/Media.php:
```diff
@@ -124,6 +124,8 @@
 
     protected $category;
 
+    protected $assetPaths = array();
```
PHPDoc missing

OskarStark commented:
thank you @xaben waiting for the rebase, the PR looks good to me

OskarStark commented on Model/MediaInterface.php:
```diff
@@ -365,4 +365,16 @@ public function getGalleryHasMedias();
+    /**
+     * @return array
+     */
+    public function getAssetPaths();
```
please remove this and create a new PR where we add these methods to the interface, otherwise its a BC break

OskarStark replied:
in the new PR, pleaase use then @inheritdoc from the interface in `Media.php`

OskarStark commented on Provider/BaseVideoProvider.php:
```diff
@@ -162,13 +162,6 @@ public function postPersist(MediaInterface $media)
     /**
-     * {@inheritdoc}
-     */
-    public function postRemove(MediaInterface $media)
```
we cant remove this method, its a BC break

OskarStark replied:
ah sorry, it uses the BaseProvider for this, i understand :+1:

OskarStark commented on Tests/Provider/VimeoProviderTest.php:
```diff
@@ -97,6 +97,11 @@ public function testThumbnail()
+
+        $provider->preRemove($media);
+        $this->assertTrue(in_array('default/0011/24/thumb_1023457_reference.jpg', $media->getAssetPaths()), 'Media::getAssetPaths() return the correct paths to delete');
+        $this->assertTrue(in_array('default/0011/24/thumb_1023457_big.jpg', $media->getAssetPaths()), 'Media::getAssetPaths() return the correct paths to delete');
```
maybe we should assert, that `$media->getAssetPaths() is equal to our testArray`, because otherwise there could be more paths in `$media->getAssetPaths()` or asssert the count == 2

xaben commented:
@OskarStark I rebased the PR, fixed the tests and removed the methods from the interface in order not to break BC. I'll create a new PR for the BC change in order to include the methods in the interface (after this gets merged).

core23 commented on Model/Media.php:
```diff
@@ -143,6 +143,8 @@
      */
     protected $category;
 
+    protected $assetPaths = array();
```
Missing PHPdoc

xaben replied:
Fixed

xaben commented:
@core23 @OskarStark Anything else that needs fixing for this PR?

core23 commented on Provider/BaseProvider.php:
```diff
@@ -186,15 +186,14 @@ public function getProviderMetadata()
+            }
         }
+
+        $media->setAssetPaths($assetPaths);
```
We create the new property only for this hack? :-1:

Soullivaneuh commented:
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.
